keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+  between multiple people/teams
+- Allows for updating the abstract policy definition separate from the concrete
+  policy
+
+Cons:
+
+- Significantly more complex implementation
+  - CRDs need to be created when a ValidatingAdmissionConfiguration is created
+  - Requires the API server establish watches on the CRs to accumulate all the
+    configuration information
+- Possible to implement using Option 1. A controller can construct a
+  resource for the API in Option 1 from custom resources that define this
+  option (e.g. via a controller).
+- CRD lifecycle prevents CRs from being written until the CRD is written, which
+  means that until a ValidatingAdmissionConfiguration is written AND a CRD is
+  created for it, no CRs can be written.  Applying multiple resources
+  (e.g. kubectl apply -f <lots of yaml files>) only works if the apply
+  operations are ordered and objects are verified to exist at specific steps.
+
+#### Additional Capabilities (WIP)
+
+In addition to the high level API structure proposed above, there are quite
+a few capatiblities that are important to this enhancement that we are working
+on defining. These are all work-in-progress.
+
+Rule Message:
+
+- message formatting (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?)
+- severity (e.g. "Warning")
+- status type equivalent (e.g. Forbidden can be returned from a webhook using
+  HTTP 403) (can a match rule override this? or can it be selected via a CEL
+  expression?)
+- Similar to how each match has configuration, can each match also select
+  the severity of messages?
+
+Rule scope:
+
+- Dereferencing to deeply nested fields in CEL can be burdonsome. A field
+  containing a field path would allow a CEL rule to be scope to some node in
+  the schema.
+- Also, the idea of named scopes (e.g. "ALL_CONTAINERS") would allow a single
+  rule to be applied to all container definitions in a resource (e.g. both
+  "containers" and "initContainers" in a pod)
+
+Failure Policy:
+
+- Not the same as webhooks, where failure policy is about the remote request
+- What if a CEL rule evaluates to an error?
+- What if a CEL rule is for a CRD kind and the CRD changes such that the CEL
+  rule fails type checking? (or the CRD does not exist?)
+
+Decision Rules:
+
+- Are overlapping matches combined using an OR or AND? Can this be configured?
+- Can a validation rule depend on another validation rule (only evaluate if the
+  other rule is true?)
+- How to make it so that an "not authorized" error prevents all other validation
+  messages from being returned (to prevent exfiltration)?
+
+Composition utilities:
+
+- Ability to define a sub-expression and then use it in validation rules (xref
+  cel-policy-template "terms")
+
+Transition rules:
+
+- Should we use "self" and "oldSelf" ? How do transition rules work for delete?
```

Comment:
as much as possible, I would encourage matching admission webhook AdmissionReview semantics. `object` is null for DELETE operations, and `oldObject` is null for `CREATE` operations

Reply:
+1, I'm in favor of using "object" as the variable throughout this KEP (instead of "self"). That keeps things consistent with AdmissionReview.  I've incorporated this.